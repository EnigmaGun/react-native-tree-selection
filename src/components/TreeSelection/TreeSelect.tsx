import React, { useCallback, useEffect, useState } from 'react';
import {
  FlatList,
  SafeAreaView,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import { dummyTreeData } from '../../constants';
import styles from './styles';
import type {
  IItems,
  ItemTypes,
  SubcategoryDataTypes,
  TreeSelectTypes,
} from './types';

const TreeSelect = ({
  treeData,
  parentOnPress = ([]) => {},
  childOnPress = (_value: {}) => {},
}: TreeSelectTypes) => {
  const [listData, setListData] = useState(treeData ?? dummyTreeData);
  const [refresh, setRefresh] = useState<boolean>(false);

  const checkRender = useCallback((data: Array<SubcategoryDataTypes>) => {
    if (!data?.hasOwnProperty('isExpanded')) {
      data.map((item: SubcategoryDataTypes) => {
        if (!item?.hasOwnProperty('id')) {
          item.isExpanded = false;
        }
      });
    }
    data.map(item => {
      item?.subcategory?.length > 0 &&
        checkRender(item?.subcategory as unknown as []);
    });
  }, []);

  useEffect(() => {
    checkRender(listData as unknown as []);
    setListData(listData);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const onPressItems = (item: IItems) => {
    parentOnPress([item?.category_name, item?.category_id]);
    item.isExpanded = !item?.isExpanded;
    setRefresh(!refresh);
  };

  const renderTree = ({ item }: ItemTypes): JSX.Element => {
    return (
      <>
        {item?.category_name && (
          <TouchableOpacity
            onPress={() => {
              onPressItems(item);
            }}
            style={styles.parentStyles}>
            <Text style={styles.text}>{item?.category_name}</Text>
          </TouchableOpacity>
        )}
        {item?.value && (
          <TouchableOpacity onPress={() => childOnPress(item)}>
            <Text style={styles.text}>{item?.value}</Text>
          </TouchableOpacity>
        )}
        <View>
          {item?.subcategory?.length > 0 && item?.isExpanded && (
            <View style={styles.innerContainer}>
              <FlatList
                data={item?.subcategory as unknown as []}
                renderItem={renderTree}
                keyExtractor={(data, index) =>
                  `${index}uniquie ID${data?.category_id}`
                }
              />
            </View>
          )}
        </View>
      </>
    );
  };

  return (
    <SafeAreaView style={styles.container}>
      <FlatList
        data={listData as unknown as []}
        renderItem={renderTree}
        keyExtractor={(data, index) => `${index}${data?.category_id}`}
        extraData={refresh}
      />
    </SafeAreaView>
  );
};

export default TreeSelect;
