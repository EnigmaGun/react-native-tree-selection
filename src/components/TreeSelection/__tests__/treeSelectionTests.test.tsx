import React from 'react';
import { render } from '@testing-library/react-native';
import TreeSelect from '../TreeSelect';

jest.useFakeTimers();

describe('Tree View component', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(<TreeSelect treeData={[]} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
