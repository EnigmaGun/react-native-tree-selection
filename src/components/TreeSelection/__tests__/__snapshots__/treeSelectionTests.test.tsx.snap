// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree View component Match Snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#022441",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    data={Array []}
    extraData={false}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</RCTSafeAreaView>
`;
